File: src/main/kotlin/net/techcable/ivan/ast/lexer.kt
package net.techcable.ivan.ast

import net.techcable.ivan.ast.TokenType.*
import kotlin.math.min

data class Span(
    val line: Int,
    val column: Int
)

data class Token(
    val tokenType: TokenType,
    val value: String,
    val span: Span
)

enum class TokenType(val text: String?) {
    Identifier(null),
    DocComment(null),

    // Symbols
    OpenBracket("{"),
    CloseBracket("}"),
    Colon(":"),
    Semicolon(";"),
    Comma(","),
    And("&"),
    Star("*"),
    OpenParen("("),
    CloseParen(")"),

    // Keywords
    SelfType("Self"),
    Self("self"),
    Interface("interface"),
    Function("fun"),
    Raw("raw"),
    Mut("mut"),
    Own("own"),
    Opaque("opaque"),
    Type("type"),
}

val keywordMap: Map<String, TokenType> = HashMap<String, TokenType>().apply {
    for (token in values()) {
        val text = token.text ?: continue
        assert(text !in this) {
            "Duplicate tokens for ${text}: ${this[text]} and $token"
        }
        this[text] = token // TODO: Duplicate checks
    }
}

fun lex(text: String): List<Token> {
    val lexer = Lexer(text)
    val result = ArrayList<Token>(text.length / 8)
    lexLoop@ while (lexer.peek() != null) {
        val c = lexer.peek()!!
        result.add(
            when (c) {
                '{' -> lexer.expectToken(OpenBracket)
                '}' -> lexer.expectToken(CloseBracket)
                ':' -> lexer.expectToken(Colon)
                ';' -> lexer.expectToken(Semicolon)
                ',' -> lexer.expectToken(Comma)
                '&' -> lexer.expectToken(And)
                '*' -> lexer.expectToken(Star)
                '(' -> lexer.expectToken(OpenParen)
                ')' -> lexer.expectToken(CloseParen)
                '/' -> {
                    when (lexer.peek(1)) {
                        '/' -> {
                            // Skip till end of line
                            while (lexer.index < lexer.text.length &&
                                lexer.text[lexer.index] != '\n'
                            ) {
                                lexer.pop()
                            }
                            continue@lexLoop
                        }
                        '*' -> {
                            if (lexer.peek(2) != '*') {
                                throw ParseException(
                                    "Block comments are unsupported",
                                    lexer.span()
                                )
                            }
                            val startSpan = lexer.span()
                            lexer.skipText("/**\n")
                            val closeIndex = lexer.text.indexOf(
                                " */", // NOTE: We require proper formatting
                                startIndex = lexer.index
                            )
                            if (closeIndex < 0) {
                                throw ParseException(
                                    "Unable to find end of comment",
                                    startSpan
                                )
                            } else {
                                val commentText = lexer.text.substring(
                                    lexer.index, closeIndex
                                )
                                lexer.skipText(commentText)
                                lexer.skipText(" */")
                                Token(
                                    tokenType = DocComment,
                                    value = commentText,
                                    span = startSpan
                                )
                            }
                        }
                        else -> throw ParseException(lexer.span())
                    }
                }
                else -> {
                    // Ignore whitespace
                    if (c.isWhitespace()) {
                        lexer.pop()
                        continue@lexLoop
                    }
                    val startSpan = lexer.span()
                    val ident = lexer.parseIdentifier()
                    val keyword = keywordMap[ident]
                    if (keyword != null) {
                        Token(
                            tokenType = keyword,
                            value = keyword.text!!,
                            span = startSpan
                        )
                    } else {
                        Token(
                            tokenType = Identifier,
                            value = ident,
                            span = startSpan
                        )
                    }
                }
            }
        )
    }
    return result
}

private class Lexer(val text: String) {
    var index = 0
        private set

    fun peek(ahead: Int = 0): Char? {
        if (this.index + ahead >= this.text.length) return null
        return this.text[this.index + ahead]
    }

    fun pop(): Char {
        if (this.index >= this.text.length) {
            throw ParseException("Unexpected EOF", this.span())
        }
        return this.text[this.index++]
    }

    fun expectToken(tokenType: TokenType): Token {
        val tokenText = tokenType.text
            ?: throw IllegalArgumentException("Expected token w/ static text: $tokenType")
        val span = this.span()
        this.skipText(tokenText)
        return Token(tokenType, tokenText, span)
    }

    fun skipText(text: String) {
        if (this.text.regionMatches(this.index, text, 0, text.length)) {
            this.index += text.length
        } else {
            val actual = this.text.substring(
                this.index, min(
                    this.index + text.length,
                    this.text.length
                )
            )
            throw ParseException("Expected $text but got $actual", span())
        }
    }

    fun parseIdentifier(): String {
        val startIndex = this.index
        val start = this.peek()
        if (start == null || !start.isJavaIdentifierStart()) {
            throw ParseException("Expected an identifier", this.span())
        }
        do {
            this.index += 1
        } while (this.peek()?.isJavaIdentifierPart() == true)
        return this.text.substring(startIndex, this.index)
    }

    private var cachedLine = -1
    private var lineStart = 0
    private var lineEnd = this.text.indexOf('\n').let { end ->
        if (end >= 0) end else this.text.length
    }
    fun span(): Span {
        if (cachedLine > 0 && this.index <= this.lineEnd) {
            return Span(
                line = this.cachedLine,
                column = this.index - this.lineStart
            )
        }
        // Fallback to compute
        var line = if (this.cachedLine > 0) this.cachedLine else 1
        var index = this.lineStart
        check(index in 0..this.index)
        while (index < this.index) {
            if (this.text[index] == '\n') {
                line += 1
                lineStart = lineEnd + 1
                val end = this.text.indexOf('\n', index + 1)
                lineEnd = if (end >= 0) end else this.text.length
            }
            index += 1
        }
        check(index >= lineStart && index <= lineEnd)
        this.cachedLine = line
        return Span(
            line = line,
            column = index - lineStart
        )
    }
}

class ParseException(msg: String, val span: Span) : RuntimeException(msg) {
    constructor(span: Span) : this("Unexpected text", span)
}

File: ivan/ast/lexer.py
from dataclasses import dataclass
from enum import Enum
from typing import Iterable, Optional


@dataclass(frozen=True)
class Span:
    """A position in the original text file."""
    __slots__ = "line", "column"
    line: int
    column: int


VALID_SYMBOLS = {"{", "}", ":", ";", ",", "&", "*", '@', '=', "(", ")"}
VALID_KEYWORDS = {"Self", "self", "interface", "fun", "raw", "mut", "own", "opaque", "type",
                  "true", "false", "opt", "field", "default"}


class TokenType(Enum):
    IDENTIFIER = 0
    DOC_COMMENT = 1
    SYMBOL = 2
    KEYWORD = 3
    STRING_LITERAL = 4

class ParseException(Exception):
    span: Span

    def __init__(self, msg: str, span: Span):
        super().__init__(msg)
        self.span = span


@dataclass(frozen=True)
class Token:
    token_type: TokenType
    value: str
    span: Span
    __slots__ = "token_type", "value", "span"

    def is_keyword(self, k: str) -> bool:
        assert k in VALID_KEYWORDS
        return self.token_type == TokenType.KEYWORD and self.value == k

    def is_symbol(self, s: str):
        assert s in VALID_SYMBOLS
        return self.token_type == TokenType.SYMBOL and self.value == s


def lex_all(s: str) -> Iterable[Token]:
    if not s:
        return
    lexer = Lexer(s)
    while lexer.index < len(lexer.text):
        token = lex_next(lexer)
        if token is not None:
            yield token


def lex_next(lexer: "Lexer") -> Optional[Token]:
    c = lexer.peek()
    assert c is not None
    if c in VALID_SYMBOLS:
        span = lexer.span()
        lexer.pop()
        return Token(TokenType.SYMBOL, c, span)
    elif c == '/':
        c = lexer.peek(1)
        if c == '/':
            # Skip till the end of the line
            while lexer.index < len(lexer.text) and\
                    lexer.text[lexer.index] != '\n':
                lexer.index += 1
            return None  # This was skipped
        elif c == '*':
            if lexer.peek(2) != '*':
                raise ParseException("Block comments are unsupported", lexer.span())
            start_span = lexer.span()
            lexer.skip_text("/**\n")
            close_index = lexer.text.find('*/', lexer.index)
            if close_index < 0:
                raise ParseException("Unable to find end of comment", start_span)
            comment_text = lexer.text[lexer.index:close_index]
            lexer.skip_text(comment_text)
            comment_text = comment_text.strip()  # TODO: Is this right?
            lexer.skip_text("*/")
            return Token(
                TokenType.DOC_COMMENT,
                comment_text.strip(),  # TODO: Should we strip the comment's whitespace?
                start_span
            )
        else:
            raise ParseException(f"Unexpected char: {c}", lexer.span)
    elif c == '"':
        start_span = lexer.span()
        lexer.skip_text('"')
        result = []
        while True:
            c = lexer.peek()
            if c == '\\':
                escape_symbol = lexer.peek(1)
                if escape_symbol == '\\':
                    result.append('\\')
                elif escape_symbol == '"':
                    result.append('"')
                else:
                    raise ParseException(f"Invalid escape: {c!r}", lexer.span())
                lexer.index += 2
            elif c == '"':
                break
            else:
                result.append(c)
                lexer.index += 1
        lexer.skip_text('"')
        return Token(TokenType.STRING_LITERAL, ''.join(result), start_span)
    else:
        if c.isspace():
            # Skip all other whitespace
            while True:
                lexer.index += 1
                c = lexer.peek()
                if c is None or not c.isspace():
                    break
            return None  # We just encountered whitespace - nothing special
        start_span = lexer.span()
        ident = lexer.parse_identifier()
        if ident in VALID_KEYWORDS:
            return Token(TokenType.KEYWORD, ident, start_span)
        else:
            return Token(TokenType.IDENTIFIER, ident, start_span)


class Lexer:
    text: str
    index: int
    __slots__ = "text", "index", "_cached_line", "_line_start", "_line_end"

    def __init__(self, text: str):
        assert text, "Blank text"
        self.text = text
        self.index = 0
        self._cached_line = None
        self._line_start = 0
        end = text.find('\n')
        if end >= 0:
            self._line_end = end
        else:
            self._line_end = len(text)

    def peek(self, ahead: int = 0) -> Optional[str]:
        try:
            return self.text[self.index + ahead]
        except IndexError:
            return None

    def pop(self) -> str:
        try:
            c = self.text[self.index]
            self.index += 1
            return c
        except IndexError:
            raise ParseException("Unexpected EOF", self.span())

    def skip_text(self, expected: str):
        start = self.index
        end = self.index + len(expected)
        actual = self.text[start:end]
        if expected == actual:
            self.index = end
        else:
            raise ParseException(f"Expected {expected!r} but got {actual!r}", self.span())

    def parse_identifier(self):
        start = self.index
        end = start
        text = self.text
        while end < len(text) and (text[end].isidentifier() or text[end].isdigit()):
            end += 1
        # NOTE: Digits can't start identifier
        if start == end or text[start].isdigit():
            actual = repr(text[start]) if start < len(text) else "EOF"
            raise ParseException(f"Expected an identifier, but got {actual}", self.span())
        else:
            self.index = end
            return self.text[start:end]

    def span(self):
        if self.index < self._line_end and\
                self._cached_line is not None:
            return Span(self._cached_line, self.index - self._line_start)
        # Fallback to compute
        line = self._cached_line if self._cached_line is not None else 1
        index = self._line_start
        assert 0 <= index <= self.index
        while index < self.index:
            if self.text[index] == '\n':
                line += 1
                self._line_start = self._line_end + 1
                end = self.text.find('\n', index + 1)
                self._line_end = end if end >= 0 else len(self.text)
            index += 1
        assert self._line_start <= index <= self._line_end
        self._cached_line = line
        return Span(line, column=index - self._line_start)

    def __repr__(self):
        return f"Lexer(index={self.index}, span={self.span()!r})"
